### PR TITLE
Mail pulls

### LINKED ISSUE #723: Pull Sonarr commit 'New: Replace SmtpClient with Mailkit'
Pull https://github.com/Sonarr/Sonarr/commit/579c4433494a848aa75abb09d78af7e98ebb96a6

```
git cherry-pick -ex 579c4433494a848aa75abb09d78af7e98ebb96a6
```

This commit has 3 conflict(s)

You can ask me to `close all`, `open pr here` or `open pr everywhere`

[AB#629](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/629)

### LINKED ISSUE #854: Pull Sonarr commit 'Debug logging for email notifications'
Pull https://github.com/Sonarr/Sonarr/commit/ab4f57f2fa65ebb51cecdbb8c120512c09c197c3

```
git cherry-pick -ex ab4f57f2fa65ebb51cecdbb8c120512c09c197c3
```

This commit has 1 conflict(s)

You can ask me to `close all`, `open pr here` or `open pr everywhere`

[AB#724](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/724)

### LINKED ISSUE #884: Pull Sonarr commit 'New: Multiple Recipients on Email Notifications (Also CC, BCC)'
Pull https://github.com/Sonarr/Sonarr/commit/01995e686dafe0933c59e48100b2f23f58afd66e

```
git cherry-pick -ex 01995e686dafe0933c59e48100b2f23f58afd66e
```

This commit has 2 conflict(s)

You can ask me to `close all`, `open pr here` or `open pr everywhere`

[AB#734](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/734)

### LINKED ISSUE #951: Pull Sonarr commit 'Refactor Email and add more logging'
Pull https://github.com/Sonarr/Sonarr/commit/5301620ecfd6cffb6014e986d1ce5b6bbaace93a

```
git cherry-pick -ex 5301620ecfd6cffb6014e986d1ce5b6bbaace93a
```

This commit has 2 conflict(s)

You can ask me to `close all`, `open pr here` or `open pr everywhere`

[AB#774](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/774)

### LINKED ISSUE #954: Pull Sonarr commit 'Set timeout for sending email'
Pull https://github.com/Sonarr/Sonarr/commit/5449389ca5295f80fe7d3e727a8f5c96abe146ae

```
git cherry-pick -ex 5449389ca5295f80fe7d3e727a8f5c96abe146ae
```

This commit has 1 conflict(s)

You can ask me to `close all`, `open pr here` or `open pr everywhere`

[AB#776](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/776)

### LINKED ISSUE #980: Pull Sonarr commit 'New: Mailgun connection'
Pull https://github.com/Sonarr/Sonarr/commit/55752a6c6213c1d83d347ba0f6870aa6c1cc0770

```
git cherry-pick -ex 55752a6c6213c1d83d347ba0f6870aa6c1cc0770
```

This commit has 0 conflict(s)

You can ask me to `close all`, `open pr here` or `open pr everywhere`

[AB#512](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/512)

### LINKED ISSUE #982: Pull Sonarr commit 'Fixed: Parsing RSS with null values'
Pull https://github.com/Sonarr/Sonarr/commit/8175f19442273e13517b2d507e7a4c33587f1b15

```
git cherry-pick -ex 8175f19442273e13517b2d507e7a4c33587f1b15
```

This commit has 1 conflict(s)

You can ask me to `close all`, `open pr here` or `open pr everywhere`

[AB#513](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/513)

### PR DESCRIPTION
**Pull request by bakerboy448**
#### Database Migration
No

#### Description
Various pulls; primarily mail related

#### Todos
- Tests
-  Wiki Updates

#### Issues Fixed or Closed by this PR

* Closes #981
* Closes #980
* Closes #982
* Fixes #854
* Fixes #884
* Fixes #951
* Fixes #954
* Closes #723
* [AB#629](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/629)
* [AB#776](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/776)
* [AB#774](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/774)
* [AB#734](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/734)
* [AB#724](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/724)
* [AB#513](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/513)
* [AB#512](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/512)

### CODE REVIEW

File: src/NzbDrone.Core.Test/Datastore/Migration/013_email_multiple_addressesFixture.cs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        public bool OnRename { get; set; }
+        public bool OnDownloadFailure { get; set; }
+        public bool OnImportFailure { get; set; }
+        public bool OnTrackRetag { get; set; }
+        public bool OnHealthIssue { get; set; }
+        public bool IncludeHealthWarnings { get; set; }
+        public List<int> Tags { get; set; }
+    }
+
+    public class EmailSettings48
```
Comment (Qstick) [inline src/NzbDrone.Core.Test/Datastore/Migration/013_email_multiple_addressesFixture.cs]: EmailSettings12

File: src/NzbDrone.Core.Test/Datastore/Migration/013_email_multiple_addressesFixture.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        public string Server { get; set; }
+        public int Port { get; set; }
+        public bool Ssl { get; set; }
+        public string Username { get; set; }
+        public string Password { get; set; }
+        public string From { get; set; }
+        public string To { get; set; }
+    }
+
+    public class EmailSettings49
```
Comment (Qstick) [inline src/NzbDrone.Core.Test/Datastore/Migration/013_email_multiple_addressesFixture.cs]: EmailSettings13

File: src/NzbDrone.Core/Datastore/Migration/013_email_multiple_addresses.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            var updateSql = "UPDATE Notifications SET Settings = @Settings WHERE Id = @Id";
+            conn.Execute(updateSql, corrected, transaction: tran);
+        }
+
+        private class ProviderDefinition166 : ModelBase
+        {
+            public string Settings { get; set; }
+        }
+
+        private class EmailSettings48
```
Comment (Qstick) [inline src/NzbDrone.Core/Datastore/Migration/013_email_multiple_addresses.cs]: EmailSettings12

File: src/NzbDrone.Core/Datastore/Migration/013_email_multiple_addresses.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            public string Server { get; set; }
+            public int Port { get; set; }
+            public bool Ssl { get; set; }
+            public string Username { get; set; }
+            public string Password { get; set; }
+            public string From { get; set; }
+            public string To { get; set; }
+        }
+
+        private class EmailSettings49
```
Comment (Qstick) [inline src/NzbDrone.Core/Datastore/Migration/013_email_multiple_addresses.cs]: EmailSettings13

File: src/NzbDrone.Core.Test/Datastore/Migration/013_email_multiple_addressesFixture.cs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        public string ConfigContract { get; set; }
+        public EmailSettings49 Settings { get; set; }
+        public string Name { get; set; }
+        public bool OnGrab { get; set; }
+        public bool OnReleaseImport { get; set; }
+        public bool OnUpgrade { get; set; }
+        public bool OnRename { get; set; }
+        public bool OnDownloadFailure { get; set; }
+        public bool OnImportFailure { get; set; }
+        public bool OnTrackRetag { get; set; }
```
Comment (Qstick) [inline src/NzbDrone.Core.Test/Datastore/Migration/013_email_multiple_addressesFixture.cs]: OnBookRetag

Ref. https://github.com/Readarr/Readarr/blob/8406dbe49de5b06e0ec8dfad46c1b2575c4dff74/src/NzbDrone.Core/Notifications/NotificationBase.cs#L63

File: src/NzbDrone.Core.Test/Datastore/Migration/013_email_multiple_addressesFixture.cs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+            var db = WithMigrationTestDb(c =>
+            {
+                c.Insert.IntoTable("Notifications").Row(new
+                {
+                    OnGrab = true,
+                    OnReleaseImport = true,
+                    OnUpgrade = true,
+                    OnDownloadFailure = true,
+                    OnImportFailure = true,
+                    OnTrackRetag = true,
```
Comment (Qstick) [inline src/NzbDrone.Core.Test/Datastore/Migration/013_email_multiple_addressesFixture.cs]: OnBookRetag

Ref. https://github.com/Readarr/Readarr/blob/8406dbe49de5b06e0ec8dfad46c1b2575c4dff74/src/NzbDrone.Core/Notifications/NotificationBase.cs#L63

File: src/NzbDrone.Core/Notifications/Email/Email.cs [hunk 7]
```diff
@@ -21,40 +27,143 @@ public override void OnGrab(GrabMessage grabMessage)
...
 
-            _emailService.SendEmail(Settings, BOOK_GRABBED_TITLE_BRANDED, body);
+            SendEmail(Settings, BOOK_GRABBED_TITLE_BRANDED, body);
         }
 
         public override void OnReleaseImport(BookDownloadMessage message)
         {
             var body = $"{message.Message} Downloaded and sorted.";
 
-            var paths = Settings.AttachFiles ? message.BookFiles.SelectList(a => a.Path) : null;
```
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Email/Email.cs:31]: We don't want to drop this

File: src/NzbDrone.Core/Datastore/Migration/013_email_multiple_addresses.cs [hunk 8]
```diff
@@ -0,0 +1,97 @@
+using System;
```
Comment (Qstick) [inline src/NzbDrone.Core/Datastore/Migration/013_email_multiple_addresses.cs]: To seems to already be a string[] on Readarr? If so we don't need the migration

Ref.
https://github.com/Readarr/Readarr/blob/8406dbe49de5b06e0ec8dfad46c1b2575c4dff74/src/NzbDrone.Core/Notifications/Email/EmailSettings.cs#L52

**Review by Qstick — CHANGES_REQUESTED** [review]

File: src/NzbDrone.Core/Notifications/Email/Email.cs [hunk 9]
```diff
@@ -30,31 +36,136 @@ public override void OnReleaseImport(BookDownloadMessage message)
 
             var paths = Settings.AttachFiles ? message.BookFiles.SelectList(a => a.Path) : null;
 
-            _emailService.SendEmail(Settings, BOOK_DOWNLOADED_TITLE_BRANDED, body, false, paths);
+            SendEmail(Settings, BOOK_DOWNLOADED_TITLE_BRANDED, body);
```
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Email/Email.cs]: You need update this to accept the attachment paths, like the old method

File: src/NzbDrone.Core/Readarr.Core.csproj [hunk 10]
```diff
@@ -13,6 +13,7 @@
     <PackageReference Include="FluentMigrator.Runner" />
     <PackageReference Include="FluentMigrator.Runner.SQLite" />
     <PackageReference Include="FluentValidation" />
+    <PackageReference Include="MailKit" Version="2.10.1" />
```
Comment (Qstick) [inline src/NzbDrone.Core/Readarr.Core.csproj]: Drop the version number to fix build.

**Review by Qstick — DISMISSED** [review]

**Comment by Qstick** [comment]
Didn’t know ta added Directory.Package.Props to centralize package versions, you need to add Malikit ref there as well

**Comment by bakerboy448** [comment]
- [x] ToDo: Figure out MailKit attachment handling and correct Email.cs as needed

File: src/NzbDrone.Core/Notifications/Email/Email.cs [hunk 11]
```diff
@@ -30,31 +36,144 @@ public override void OnReleaseImport(BookDownloadMessage message)
...
+            email.Body = new TextPart(htmlBody ? "html" : "plain")
+            {
+                Text = body
+            };
+
+            if (attachmentUrls != null)
+            {
+                foreach (var url in attachmentUrls)
+                {
+                    email.Attachments.Add(new MimePart(url));
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Notifications/Email/Email.cs]: specifically; figure this out

Comment (bakerboy448) [reply src/NzbDrone.Core/Notifications/Email/Email.cs]: http://www.mimekit.net/docs/html/P_MimeKit_MimeMessage_Attachments.htm
and
http://www.mimekit.net/docs/html/Creating-Messages.htm

**Comment by bakerboy448** [comment]
Attachments work; tested locally with blackhole for an import.

We may want to somehow guard and set a size limit?

**Comment by RetroDadson** [comment]
Take a look at this, I think we should try and guard on size by checking the maximum size from a capabilities check 

http://www.mimekit.net/docs/html/P_MailKit_Net_Smtp_SmtpClient_MaxSize.htm

**Comment by bakerboy448** [comment]
> Take a look at this, I think we should try and guard on size by checking the maximum size from a capabilities check
> 
> http://www.mimekit.net/docs/html/P_MailKit_Net_Smtp_SmtpClient_MaxSize.htm

agreed  we'd also need to know the proposed attachment(s) sizes to check against

How should we handle a multiple part book where only 1 of the parts is under the size limit?

(also way out of my depth here)

**Comment by RetroDadson** [comment]
We shouldn't be handling it on a per attachment basis but on the size of the whole mail. I'll try have a look later today

**Review by Qstick — APPROVED** [review]

**Pull request merged**